### ISSUE TITLE
Language extensions and warning flags

### ISSUE THREAD

**Issue by simonpj** [ghc-proposals/ghc-proposals#615]
Richard and I had a conversation yesterday that uncovered a difference of thinking
that might be worth making explicit, since it lurks in the background of more specific
conversations.

Consider multi-parameter type classes.  We could treat this feature in two ways:

1. (The status quo.)  As a language extension `-XMultiParamTypeClasses`.  If you use a multi-param type class without saying `-XMultiParamTypeClasses` we cause an error, but with an informal warning: "Maybe you meant to use MultiParamTypeClasses".

2. (Possible alternative.) GHC always understands multi-parameter classes, but can warn or error if you use them:
   * `-Werrror=multi-param-type-classes`: error if you use them (this is the default)
   * `-Wwarn=multi-param-type-classes`: warn if you use them.
   * `-Wno-multi-param-type-classes`: fully allow multi-param type classes, no errors, no warnings.
   * `-XMultiParamTypeClasses` is merely a synonym for `-Wno-multi-param-type-classes`

Richard argues that the second approach is better. I'll let him say why!  One small point is that we have no equivalent today for `-Wwarn=multi-param-type-classes`, and (thinking about all langauge extensions) that might be a nice thing to have.

But at least I think it's clear that the current situation is inconsistent.  Consider `-Wname-shadowing`.  Suppose Haskell had prohibited shadowing from Day 1.  Then someone says "but I want shadowing; give me an extension". Then, with our current conventions, we'd get `-XAllowNameShadowing` as a langauge extension.  Conclusion: the current split between warnings and language extensions is really a historical accident, not a principled choice.

Some language extensions *change the behaviour of an existing feature*.  (Even now I'm having a hard time enumerating exactly which those are; but perhaps `OverloadedRecordDot` is an example, because without it `x.y` means function composition.  And `RebindableSyntax`.)  Extensions that change behaviour really do need to be language extensions; the warning approach won't cut it.   But arguably it'd be good to know which extensions *change* behavior, and which just *add* behaviour.

I think Richard would like us to move, as a matter of policy, towards warning-oriented flags for extrensions that add behaviour.  I'm not arguing a position on this, but I wanted to articulate the choice explicitly.

**Comment by AntC2** [comment]
Thank you @simonpj & @goldfirere, a stimulating approach. And I've seen Richard float the idea in a few places already. A couple of fishhooks (sorry).

* `-XOverlappingInstances` we used to have a module-wide extension. We now have pragmas within the instance `{-# OVERLAPPING, OVERLAPPABLE, OVERLAPS, INCOHERENT #-}`. In removing the module-wide extensions, there's now awkwardness for program analyser tools in detecting which settings are in force -- particularly critical for `INCOHERENT`.

* `-XScopedTypeVariables` (which is driving some of these concerns -- it's exactly " _change the behaviour of an existing feature_") is particularly unruly: it doesn't introduce any syntax; rather it changes the meaning of explicit `forall` in a signature; in itself `forall` appearing shouldn't trigger a warning/error -- it's only if the introduced tyvars appear in type annotations within the function's decl. (Richard had previously suggested a flag like `-XExtendedForallScope`.) A hypothetical `-Wwarn=extended-forall-scope` should (?) detect if a `forall`'d tyvar is used in a annotation within the function decl.

**Comment by goldfirere** [comment]
Thanks, SImon, for writing this up. I have been wondering in this direction for a while. To be explicit, I'm not fully confident that "let's just change extensions to warnings" is a great idea, but I tend to think it would be an improvement. Here are my reasons (numbered in case anyone wants to refer to them):

1. As @simonpj said, the current arrangement of extensions-vs-warnings is unprincipled. And it's a real mess in places. As examples: `-XDeriveFunctor` just allows new code, with no possibility of warning; `-fdefer-type-errors` allows new code, but you also have to say `-Wno-deferred-type-errors` to squelch the warning; `-XPartialTypeSignatures` allows new code, but you also have to say `-Wno-partial-type-signatures` to squelch the warning; missing methods are already allowed, but you can stop them with `-Werror=missing-methods`. To me, all these different features are on similar footing and should be controlled by the same mechanism.

2. For better or worse, the standards process for Haskell has stopped. When there was a standard, portable Haskell language, it made sense to compile that language, and then offer opportunities to go beyond the standard. However, without a standards process, the motivation against just adding new language features is much weaker.

3. No one else (that I know of) does what we do regarding language extensions -- and we struggle with the weight of the extensions. (That is, the fact that no one else does something is not a reason we shouldn't, unless we also find it troublesome.) That is, when Java or C++ comes out with a brand new feature (and it's always backward compatible), it's not gated behind any flag. It's just a new feature of the language. I don't see a compelling reason not to do the same in Haskell. (Important exception: when a new feature changes the reasoning properties of Haskell programs in some way, that might be a reason for a language extension.)

Just imagine a world where the majority of our extensions were just warnings -- or not even that. (For example, would anyone enable `-Wderive-functor`? I doubt it.) All of these debates about extensions would be simpler. We would be left only with those features that actually change the behavior of Haskell programs, a much smaller set than we deal with today.

**Comment by simonpj** [comment]
> we struggle with the weight of the extensions

Yes -- but it's not clear to me that reframing each "extension" as a "warning" changes the weight involved. Why would "all these debates about extensions be simpler"?  If you want to make that case, by all means do so, but currently you are implying that the weight is less, without explicilty making, or justifying, that claim.

**Comment by Ericson2314** [comment]
I think "monotonic" extensions are extremely good in conjunction with https://github.com/ghc-proposals/ghc-proposals/pull/601 and being clear about what is stable. We can't just say everything to get released is instant-stable.

But once we explicit decide an extension is stable, yes we have an opportunity to just incorporate as part of a stable language version (or edition in the Rust parlance) and get rid the language extension flag.

**Comment by georgefst** [comment]
Previous discussion: #502.

**Comment by simonmar** [comment]
I really like the idea of using the warning system for extensions, however my worry is that because we can't use it universally for all extensions we'd end up with a different kind of inconsistency. There are a extensions that change behaviour, and there are a dozen or so extensions that steal existing syntax and would break existing programs (so they can't be migrated to warnings). I took a quick look through the lexer and parser and came up with the following list, which probably isn't exhaustive:

Stolen syntax extensions:
* PatternSynonyms (steals `pattern`)
* TypeApplications (steals `@`)
* BinaryLiterals (steals `0b10101`)
* NegativeLiterals (not sure about this one)
* RecursiveDo (steals `mdo`, `rec`)
* Arrows (steals `rec`, `proc`)
* StaticPointers (steals `static`)
* UnboxedTuples (steals `(#)`)
* MagicHash (steals `123#`)
* TemplateHaskell (steals `[e|`)
* UnboxedSums (steals `(#)`)

**Comment by Ericson2314** [comment]
Very similar to what @simonmar is saying, sometimes we have two extensions which individually don't steal syntax, but together too. In particular in the latest versions of #608 I discuss how implicit binding with and without cross-namespace resolution (part of `-XRequiredTypeArguments` from accepted proposal #281) are two mutually incompatable designs

We can think about this order theoretrically; say extension set a <= extension set b. Then we have multiple extension which are local maxima but are globally unordered (they steal syntax from each other). All other extensions sets can be recovered by warnings (removing things from the set), but we need at least those. Additionally, I think when two maximal sets have a unique greatest lower bound (lattice meet), this is a "point of consensus" (folks agree up to there, but then the design space forks in two irreconcilable directions) we ought to have enough extensions to uniquely refer to that, rather than needing to do *either* (`-XFoo -Werror=bar` or `-XBaz -Werror=quux` which just happen to mean the same thing.)

**Comment by phadej** [comment]
> That is, when Java or C++ comes out with a brand new feature (and it's always backward compatible), it's not gated behind any flag. 

This is untrue. C++ has defined standards, and if I want to use `-std=cxx03` I can even with the latest version of compiler. (And if I try to use some new feature, it won't work).

> let's just change extensions to warnings

> However, without a standards process, the motivation against just adding new language features is much weaker.

This also Will **require** a way to specify in `.cabal` files the GHC **version range**. Currently there isn't. We either use `base` version as a proxy, or (better option when it's works) require compiler functionality with e.g. `other-extensions: QuantifiedConstraints`

**Comment by konsumlamm** [comment]
Using warnings for language extensions feels wrong. Warnings are supposed to warn you about potentially wrong/bad/dangerous code, but I can't see anything about `MultiParamTypeClasses` or `DeriveFunctor` that would be worth warning about, other than it not being part of the standard. I could understand it for something like `UndecidableInstances` though.

**Comment by simonpj** [comment]
> Using warnings for language extensions feels wrong.

It's a matter of perspective.  I think Ricahrd would say "using language extensions for warnings (e.g. warning: I undersand multiparameter type classes, but you asked me to warn when you used one) feels wrong".

------------------------------

Stepping back, there is a fork in the road, described in the inital post on this thread:
1. Do we double down on language extensions?  If so, then in the "missing-method" debate (#571) we would not add a warning flag; instead we would add `-XAllowMissingMethods`, which is on by default.  Then in a later language edition we might switch it off by default.
2. Do we move (as rapidly as possible) towards warnings?  If so tehn in #571 we would use warnings; and perhaps we would change exixting language extensions into warnings.

Personally I favour (1), for two reasons:
* (1) is the status quo.  We have hundreds of language extensions, many of which should be warning under plan (2).  So either we get a lot of inconsistency for ever, or we endure a lot of churn.  Changing the status quo has costs.
* We are agreed that under (2) some language extensions remain, because they change behaviour of existing programs.  But 
  * It means that `-GHC2024` would mean "a bunch of language extensions and warning flags" rather than just "a bunch of language extensions".  That adds conceptual complexity and some people outright don't like it (see above).
  * It's a narrow distinction; e.g. you might change only the behaviour a program that used `forall` as the name of a type variable, rather a corner case.
  * John says *"Sometimes we have two extensions which individually don't steal syntax, but together too. In particular in the latest versions of #608 I discuss how implicit binding with and without cross-namespace resolution (part of -XRequiredTypeArguments from accepted proposal #281 are two mutually incompatable designs"*.  I have not fully understood this, but if he's right that would make the distinction even harder to discern.

**Comment by nomeata** [comment]
I agree with Simon. I am very sympathetic to the idea of framing former extensions that have become part of the modern Haskell as something you can ask the compiler to warn or error you about, and I think it’s very useful to appreciate how almost equivalent the two notions are. But I fear that actually implementing this new perspective in any meaningful way is going to be to costly. For example, while I have no issues with language editions also affecting which warnings are errors, the energy it costs to explain and convince is not worth it – precisely because `NoFeatureSomePeopleDon'tLike` and `-Werror=feature-some-people-don't-like` are so similar.

(JFTR: For me, this “doubling down” does _not_ imply doubling-down on giving users the maximum choice feasible, and still think that some extensions, like `NoDeriveFunctor`, shouldn’t be available to anyone who isn’t trying to compile dusted code. But that’s a different discussion.)

**Comment by Ericson2314** [comment]
> have not fully understood this, but if he's right that would make the distinction even harder to discern.

At the risk of saying even more incomprehensible things: consider these twos situation:

1. There is one single notion of implicit binding. A user that doesn't want it perhaps is being obstinate, like a user that is doing `-XNoDeriveFunctor`. Maybe both `-XNo....` can be relegated to being mere warnings.
2. There are two notions of implicit binding which do not agree (i.e. they steal syntax from each other, i.e. they deem the *same* program valid but with two *different* semantics). A user that sees a lack of concensus around implicit binding and therefore doesn't want it I don't think is being so obstinate. I don't think they should need to use warnings to avoid implicit bindings.

**Comment by simonmar** [comment]
+1 on not moving to warnings, I don't see a clear end point that's better than what we currently have. And yes to having a language extension for `AllowMissingMethods` (and `AllowMissingFields` I guess).

**Comment by nomeata** [comment]
Shall we close this to channel further discussion to the proposal at #620?
